### ISSUE TITLE
Remove trailing lines settings go one EOL too far.

### ISSUE THREAD

**Issue by kotp** [zlovatt/obsidian-trim-whitespace#25]
This is plugin seems to be removing the last EOL from what should be the last line[^1], but ends up with the last line followed by text after the last line.  When running scripts that concatenate files, it causes "run-in" errors.

[^1]: https://eslint.org/docs/latest/rules/eol-last

**Comment by zlovatt** [comment]
Can you provide some sample markdown to test with?

**Comment by kotp** [comment]
I can, and will, but I now do not believe it is this plugin causing the problem, but a co-editor using vscode to edit the files, and vscode not knowing what a "line" is.

[single_line.md](https://github.com/zlovatt/obsidian-trim-whitespace/files/12772696/single_line.md)
[single_line.md](https://github.com/zlovatt/obsidian-trim-whitespace/files/12772699/single_line.md)

One without a line, the other with a line, both with the text "This is a single line."

**Comment by kotp** [comment]
It seems to be the plugin, just had to close and open the vault and it is doing the same thing of removing the EOL (end of line) character from what should be the last line.

**Comment by zlovatt** [comment]
This sounds like intended behaviour with the "Trim Trailing Lines" preference enabled, which removes empty lines at the end of the document.

So, one of these documents includes a linebreak after the text; disabling that pref preserves it, which seems to be the goal?

**Comment by kotp** [comment]
The last line, by definition of a line, should have an "End of Line" character.  Extraneous ones should be removed, definitely.

So if you have a collection of lines, where you have 0 or more characters followed by an EOL character, and then the last line is really 4 EOL characters, you would expect it to trim the 3 extraneous EOL characters, but preserve the last line as a line, rather than make what was the last line, "extra characters after the last line".

Good:

```plain
Original second  to the last line.\n
Last line\n
```

Not good:

```plain
Original second line, but now the last line.\n
Original last line, but now extra characters after the last line.
```

**Comment by kotp** [comment]
About changing the preferences, yes, indeed, I can "not use the plugin", that would solve the problem altogether.  But the plugin is convenient and from what I can see (not a javascript dev) pretty well designed.  I would rather it do the Right Thing™ instead, as it concerns lines.  To get around it, for sure, I have disabled the setting for the meantime to not trim at the end of the file, and just have the team pay attention to what they are doing.

**Comment by kotp** [comment]
No, I do not imply.  I missed a word that should have resulted in "by _some_ definition" . I simply missed typing a word, it should have been "by _some_ definition" and your link confirms that practice that by some definition it is a terminator, by another definition it is a separator.  This has been going on for at least 40 years, and I am well aware, so would not imply some universal rule.  It would be knowingly dishonest.

Sure, I am not sure I called it a defect, but we can go with "enhancement" if you like.  However, the description of the options state:

![image](https://github.com/zlovatt/obsidian-trim-whitespace/assets/16747/63579b4a-1475-43cf-baed-0c6b9da62393)

We can see in the description that it should "trim empty lines" not "trim the EOL from the last line".  It surely trims 'non-empty' lines as well.

I would be fine with a description change, but would be better with leaving the terminator there, with an option to use it as a separator instead of terminator instead.

You bring up a great point about being cautious with "scripts that concatenate files".  We can even expand that to say "scripts that remove things from files" as we can see this is such a script that does not do what it says it does by description.  I generally recommend caution with any script.  It is why I am looking at this plugin in the first place!

**Comment by kotp** [comment]
Since I did not also provide a link, and since we are working with plain text, and POSIX is a thing, this may be an interesting [answer] for context, including the validity of the "separator" camp.

[answer]: https://stackoverflow.com/a/729795/485864

**Comment by zlovatt** [comment]
With this code sample:
```md
Original second  to the last line.¬
Last line¬
```

You have some content ("Original second...Last line"), followed by trailing whitespace (the new line created by `\n`, stylized as `¬`).

The preference "Trim trailing lines" will remove the empty line at the end of your document, resulting in the same content, but with that trailing newline character removed.
```md
Original second  to the last line.¬
Last line
```

This is by design, and it's doing exactly what it's meant to. It does indeed conflict with the dev-centric practice of inserting a final newline at the end of the file, as these concepts are incompatible without further controls. 

---

To that end, as @velebit mentioned, it sounds like this ticket raises a feature request to have a separate preference for whether _preserve_ a single trailing newline character.

For reference, this is how vscode handles it; it uses two separate preferences, which operate independently:
1. Insert Final Newline (whether to add that extra line at the end), and
2. Trim Final Newlines (which will trim all newlines, including the final newline **unless the preference above is enabled**)

While this plugin won't ever insert that character, we can definitely add a preference to preserve it if it's there!

**Comment by kotp** [comment]
> The preference "Trim trailing lines" will remove the empty line at the end of your document, resulting in the same content, but with that trailing newline character removed.

This is described as we might be able to see in the first option that I submitted as a "screenshot", as "trim empty lines at the end of the document".

The second line is not an empty line, it has content and a newline at the end of it.

This:

```md
\n
\n
```
would be two empty lines, since there is nothing before the `newline` character.  Those I would expect to be trimmed, by the description of "trim empty lines at the end of the document".

I can not understand how this:

```md
something\n
```

would be described as an empty line at the end of a document.

I do know that some tools present a visual blank line after that, though.

> While this plugin won't ever insert that character, we can definitely add a preference to preserve it if it's there!

YAY!  That would be nice.  And definitely good news.  I would not expect things to be inserted in my documents, especially in a tool called "trim" not "insert".

**Comment by kotp** [comment]
> This sounds like intended behaviour with the "Trim Trailing Lines" preference enabled, which removes empty lines at the end of the document.
> 
> So, one of these documents includes a linebreak after the text; disabling that pref preserves it, which seems to be the goal?

Generally speaking, rather than technically speaking, preserving it, if it is there, is good, yes. 

I think you will find that what is there is the `newline` character though.  Had I presented it as Windows file, it would be `linefeed` and `carriage return` and as Mac/OS, I would have presented it as `linefeed` only.  Whatever the character(s) are that are being addressed and trimmed, as described in the settings/options of the tool/plugin, should only happen on empty lines, as it states.

Disabling that preference, though, also leaves the extras blank lines after the last non-empty line.  Ideally, I would only want blank lines to be trimmed as the description for the option states.

**Comment by husjon** [comment]
Hi, just to inform that I'm currently looking into this and will be making it configurable as part of #28.

**Comment by husjon** [comment]
Hi, just to let you know that this feature has been added as part of version 0.4.0, the release is now available.

You'll find the new option called **Max lines to keep** and is available after enabling **Trim Trailing Lines**.
